tools/sklearn/dict_h5_pickler.py

```diff
@@ -0,0 +1,510 @@
+        return tuple(obj)
+
+    def load_set(self, data):
+        obj = self.load_all(data)
+        return set(obj)
+
+    def load_dict(self, data):
+        newdict = {}
+        _keys = data[_KEYS]
+        for k in _keys:
+            try:
+                v = data[k]
+            # JSON dumps non-string key to string
+            except KeyError:
+                v = data[str(k)]
+            newdict[k] = self.load_all(v)
+        return newdict
+
+    def load_global(self, data):
+        module = data[0]
+        name = data[1]
+        func = SafePickler.find_class(module, name)
+        self.memoize(func)
+        return func
+
+    def load_reduce(self, data):
+        """
+        Build object
+        """
+        _func = data[_FUNC]
+        func = self.load_all(_func)
+        assert callable(func), "%r" % func
+
+        _args = data[_ARGS][_TUPLE]
+        args = tuple(self.load_all(_args))
+
+        try:
+            obj = args[0].__new__(args[0], * args)
+        except:
+            obj = func(*args)
+
+        _state = data.get(_STATE)
+        if _state:
+            state = self.load_all(_state)
+            setstate = getattr(obj, "__setstate__", None)
+            if setstate:
+                setstate(state)
+            else:
+                assert (type(state) is dict)
+                for k, v in state.items():
+                    setattr(obj, k, v)
+
+        self.memoize(obj)
+        return obj
+
+    def load_np_ndarray(self, data):
+        _dtype = self.load_all(data[_DTYPE])
+        _values = self.load_all(data[_VALUES])
+        obj = numpy.array(_values, dtype=_dtype)
+        return obj
+
+    def load_np_datatype(self, data):
+        _datatype = self.load_all(data[_DATATYPE])
+        _value = self.load_all(data[_VALUE])
+        obj = _datatype(_value)
+        return obj
+
+
+class H5Model:
+    """
+    Convert the dictionary model to h5 and reverse
+    """
+
+    def _create_dataset(self, file_obj, key, value):
+        """
+        Create dataset
+        """
+        try:
+            file_obj.create_dataset(key, data=json.dumps(value))
+        except:
+            file_obj.create_dataset(key, data=value)
+
+    def save_model(self, model, file_name):
+        """
+        Save the dictionary model to h5 file
+        """
+        serialised_model = ModelToDict().dump(model)
+        h5file = h5py.File(file_name, 'w')
+
+        # nested method for recursion
+        def recursive_save_model(h5file_obj, dictionary):
+            for model_key, model_value in dictionary.items():
+                type_name = type(model_value).__name__
+                try:
+                    if type_name in ['list']:
+                        if len(model_value) > 0:
+                            list_obj = all(isinstance(x, dict) for x in model_value)
+                            if list_obj is False:
+                                self._create_dataset(h5file_obj, model_key, model_value)
```

Comment:
It seems h5 can store nested list directly. Great! How about storing nested tuple? I guess it can handle that as well.

Reply:
For nested lists, I created dictionary recursively with indices as the keys and while unpacking, I remove those keys to recreate the list. It also explains the why some keys are numeric, others not